cleverhans/experimental/certification/dual_formulation.py

```diff
@@ -35,12 +38,7 @@ def __init__(self, dual_var, neural_net_param_object, test_input, true_class,
     self.input_minval = tf.convert_to_tensor(input_minval, dtype=tf.float32)
     self.input_maxval = tf.convert_to_tensor(input_maxval, dtype=tf.float32)
     self.epsilon = tf.convert_to_tensor(epsilon, dtype=tf.float32)
-    self.final_linear = (self.nn_params.final_weights[adv_class, :]
-                         - self.nn_params.final_weights[true_class, :])
-    self.final_linear = tf.reshape(self.final_linear,
-                                   shape=[tf.size(self.final_linear), 1])
-    self.final_constant = (self.nn_params.final_bias[adv_class]
-                           - self.nn_params.final_bias[true_class])
+    self.final_linear, self.final_constant = final_linear_and_constant(self.nn_params, adv_class, true_class)
```

Comment:
This caches these graph nodes, so multiple DualFormulation objects with the same arguments don't build multiple copies of final_linear / final_constant

Reply:
Currently line 49 fails (it creates a new graph node). I haven't yet investigated whether that line is feasible to cache.